ably-ios/ARTEventEmitter.h

```diff
@@ -9,14 +9,102 @@
 #import <Foundation/Foundation.h>
 #import "ARTTypes.h"
 
-@protocol ARTSubscription;
-
 @class ARTRealtime;
 
-@interface ARTEventEmitter : NSObject
+ART_ASSUME_NONNULL_BEGIN
+
+@interface __GENERIC(ARTEventListener, ItemType) : NSObject
+
+- (void)call:(ItemType)argument;
+
+@end
+
+@interface __GENERIC(ARTEventEmitter, EventType, ItemType) : NSObject
+
+- (__GENERIC(ARTEventListener, ItemType) *)on:(EventType)event call:(void (^)(ItemType))cb;
+- (void)on:(EventType)event callListener:(__GENERIC(ARTEventListener, ItemType) *)listener;
 
-- (instancetype)initWithRealtime:(ARTRealtime *)realtime;
-- (id<ARTSubscription>)on:(ARTRealtimeConnectionStateCb)cb;
-- (void)removeEvents;
+- (__GENERIC(ARTEventListener, ItemType) *)onAll:(void (^)(ItemType))cb;
+- (void)onAllCallListener:(__GENERIC(ARTEventListener, ItemType) *)listener;
+
+- (__GENERIC(ARTEventListener, ItemType) *)once:(EventType)event call:(void (^)(ItemType))cb;
+- (void)once:(EventType)event callListener:(__GENERIC(ARTEventListener, ItemType) *)listener;
+
+- (__GENERIC(ARTEventListener, ItemType) *)onceAll:(void (^)(ItemType))cb;
+- (void)onceAllCallListener:(__GENERIC(ARTEventListener, ItemType) *)listener;
+
+- (void)off:(EventType)event listener:(__GENERIC(ARTEventListener, ItemType) *)listener;
+- (void)offAll:(__GENERIC(ARTEventListener, ItemType) *)listener;
+
+- (void)emit:(EventType)event with:(ItemType)data;
 
 @end
+
+// This macro adds methods to a class header file that mimic the API of an event emitter.
+// This way you can automatically "implement the EventEmitter pattern" for a class
+// as the spec say. It's supposed to be used together with ART_EMBED_IMPLEMENTATION_EVENT_EMITTER
+// in the implementation of the class.
+#define ART_EMBED_INTERFACE_EVENT_EMITTER(EventType, ItemType) - (__GENERIC(ARTEventListener, ItemType) *)on:(EventType)event call:(void (^)(ItemType))cb;\
```

Comment:
Can protocols implement methods and have properties? I think they were just interfaces that implementors had to, well, implement themselves.